File: mcp_server_development/python/src/session15/rate_limit.py
"""レート制限（トークンバケット・Python 版）

時刻を callable で受け取ります。テストで実時間を待たないためです。
"""

from __future__ import annotations

import math
from collections.abc import Callable
from dataclasses import dataclass

DEFAULT_MAX_KEYS = 1000


@dataclass(frozen=True)
class RateLimitDecision:
    allowed: bool
    remaining: int = 0
    retry_after_ms: int = 0


class RateLimiter:
    def __init__(
        self,
        *,
        capacity: int,
        refill_per_second: float,
        now: Callable[[], int],
        max_keys: int = DEFAULT_MAX_KEYS,
    ) -> None:
        self._capacity = capacity
        self._refill = refill_per_second
        self._now = now
        self._max_keys = max_keys
        #: キー -> (残トークン, 最終更新時刻)
        self._buckets: dict[str, tuple[float, int]] = {}

    def size(self) -> int:
        return len(self._buckets)

    def try_consume(self, key: str, cost: int = 1) -> RateLimitDecision:
        at = self._now()
        current = self._buckets.get(key)
        if current is None and len(self._buckets) >= self._max_keys:
            oldest = min(self._buckets, key=lambda name: self._buckets[name][1])
            del self._buckets[oldest]

        tokens, updated_at = current if current is not None else (float(self._capacity), at)
        elapsed_ms = max(0, at - updated_at)
        tokens = min(float(self._capacity), tokens + (elapsed_ms / 1000) * self._refill)

        if tokens < cost:
            self._buckets[key] = (tokens, at)
            retry_after = math.ceil((cost - tokens) / self._refill * 1000)
            return RateLimitDecision(False, retry_after_ms=retry_after)

        self._buckets[key] = (tokens - cost, at)
        return RateLimitDecision(True, remaining=int(tokens - cost))

File: mcp_server_development/node/src/session15/rate-limit.ts
/**
 * レート制限（トークンバケット）
 *
 * 時刻を関数で受け取るのが要点です。内部で Date.now() を呼ぶと、テストが
 * 「1 秒待つ」ことになり、遅くて不安定（CI で稀に落ちる）になります。
 * 時計を注入すれば、実時間を 1 ミリ秒も待たずに「2.5 秒後」を試せます。
 */
export type RateLimitDecision =
  | { readonly allowed: true; readonly remaining: number }
  | { readonly allowed: false; readonly retryAfterMs: number };

export type RateLimiterOptions = {
  /** バケットの容量（＝瞬間的に許す最大回数） */
  readonly capacity: number;
  /** 1 秒あたりの回復量（＝定常状態で許す毎秒の回数） */
  readonly refillPerSecond: number;
  readonly now: () => number;
  /** 保持するキーの上限。無制限にするとメモリ枯渇の的になる */
  readonly maxKeys?: number;
};

export type RateLimiter = {
  tryConsume(key: string, cost?: number): RateLimitDecision;
  size(): number;
};

type Bucket = { tokens: number; updatedAt: number };

export const DEFAULT_MAX_KEYS = 1000;

export function createRateLimiter(options: RateLimiterOptions): RateLimiter {
  const { capacity, refillPerSecond, now } = options;
  const maxKeys = options.maxKeys ?? DEFAULT_MAX_KEYS;
  const buckets = new Map<string, Bucket>();

  /** 最も古いバケットを 1 つ捨てる。実運用では LRU か外部ストア（Redis 等）に置く */
  function evictOldest(): void {
    let oldestKey: string | undefined;
    let oldestAt = Number.POSITIVE_INFINITY;
    for (const [key, bucket] of buckets) {
      if (bucket.updatedAt < oldestAt) {
        oldestAt = bucket.updatedAt;
        oldestKey = key;
      }
    }
    if (oldestKey !== undefined) {
      buckets.delete(oldestKey);
    }
  }

  function tryConsume(key: string, cost = 1): RateLimitDecision {
    const at = now();
    const current = buckets.get(key);
    if (current === undefined && buckets.size >= maxKeys) {
      evictOldest();
    }
    const bucket = current ?? { tokens: capacity, updatedAt: at };
    const elapsedMs = Math.max(0, at - bucket.updatedAt);
    const tokens = Math.min(capacity, bucket.tokens + (elapsedMs / 1000) * refillPerSecond);

    if (tokens < cost) {
      // 残量は更新しておく（次回の判定で二重に回復させないため）
      buckets.set(key, { tokens, updatedAt: at });
      return { allowed: false, retryAfterMs: Math.ceil(((cost - tokens) / refillPerSecond) * 1000) };
    }
    buckets.set(key, { tokens: tokens - cost, updatedAt: at });
    return { allowed: true, remaining: Math.floor(tokens - cost) };
  }

  return { tryConsume, size: () => buckets.size };
}
